test/opt/test_fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,927 @@
+  std::vector<std::string> original;
+  // Expected cosntants after folding.
+  std::vector<std::string> expected;
+};
+
+using FoldSpecConstantOpAndCompositePassTest = PassTest<
+    ::testing::TestWithParam<FoldSpecConstantOpAndCompositePassTestCase>>;
+
+TEST_P(FoldSpecConstantOpAndCompositePassTest, ParamTestCase) {
+  AssemblyBuilder test_code_builder, expected_code_builder;
+  auto& tc = GetParam();
+  test_code_builder.AppendTypesConstantsGlobals(CommonTypesAndConstants);
+  test_code_builder.AppendTypesConstantsGlobals(tc.original);
+  expected_code_builder.AppendTypesConstantsGlobals(CommonTypesAndConstants);
+  expected_code_builder.AppendTypesConstantsGlobals(tc.expected);
+  std::string original = test_code_builder.GetCode();
+  std::string expected = expected_code_builder.GetCode();
+
+  // Run the optimization and get the result code in disassembly.
+  std::string optimized;
+  bool modified = false;
+  std::tie(optimized, modified) =
+      SinglePassRunAndDisassemble<opt::FoldSpecConstantOpAndCompositePass>(
+          original, /* skip_nop = */ true);
+
+  // Check the optimized code, but ignore the OpName instructions.
+  EXPECT_EQ(
+      StripOpNameInstructions(expected) != StripOpNameInstructions(original),
+      modified);
+  EXPECT_EQ(StripOpNameInstructions(expected),
+            StripOpNameInstructions(optimized));
+}
+
+// Tests that OpSpecConstantComposite opcodes are replace with
+// OpConstantComposite correctly.
+INSTANTIATE_TEST_CASE_P(
+    CompositeVector, FoldSpecConstantOpAndCompositePassTest,
+    ::testing::ValuesIn(std::vector<
+                        FoldSpecConstantOpAndCompositePassTestCase>({
+        // clang-format off
+            {
+              // original
+              {
+                "%spec_v2bool = OpSpecConstantComposite %v2bool %bool_true %bool_false",
+                "%spec_v2uint = OpSpecConstantComposite %v2uint %unsigned_one %unsigned_one",
+                "%spec_v2int_a = OpSpecConstantComposite %v2int %signed_one %signed_two",
+                // Spec constants whose value can not be fully resolved should
+                // not be processed.
+                "%spec_int = OpSpecConstant %int 99",
+                // "%spec_v2int_b = OpSpecConstantComposite %v2int %signed_one %spec_int",
+              },
+              // expected
+              {
+                "%spec_v2bool = OpConstantComposite %v2bool %bool_true %bool_false",
+                "%spec_v2uint = OpConstantComposite %v2uint %unsigned_one %unsigned_one",
+                "%spec_v2int_a = OpConstantComposite %v2int %signed_one %signed_two",
+                "%spec_int = OpSpecConstant %int 99",
+                // "%spec_v2int_b = OpSpecConstantComposite %v2int %signed_one %spec_int",
+              },
+            },
+        // clang-format on
+    })));
+
+// Tests for the casting across both scalar and vector bool <-> int <-> uint.
+// int 32 <-> int 64 are not included (OpSConvert is not covered).
```

Comment:
If we are not gonna cover this case, it's better to put a TODO here. Also for `OpFConvert`.